add minigun module for modsuits

Squishypone opened the pull request:
## About The Pull Request
This will add a mod module that lets you use a minigun while you're in a mod suit and also give a sound effect when you bring out and put away a minigun. The mod suit minigun uses a little bit of the Modsuit power with each shot and some power depending on how hot it is. The mod suit minigun also has a little less heat capacity compared to the backpack one
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage 
maintainers from merging your PR! -->
![dreamseeker_lWkfJbmu0L](https://user-images.githubusercontent.com/13859720/235521945-09288dc6-398e-4868-8e7d-5c330e17cdd6.png)
![dreamseeker_u3zKbfknSk](https://user-images.githubusercontent.com/13859720/235522412-8004692d-7730-4cb2-8cf9-b24a0975eaa9.gif)
![dreamseeker_gPll2BMKyK](https://user-images.githubusercontent.com/13859720/235526076-27d98ef3-9e21-4530-8b2b-a2b2d6fa27ca.gif)

Keep in mind this is really my first time adding something to the game. So if there's anything you need to change just let me know and I'll do it right away

## Why It's Good For The Game
The minigun was a really fun weapon to use especially if you're like a nuke op or death squad but with the introductions of modsuits It made it impossible to use the minigun while also in a Mod Suit With this now you can have them all the armor and stuff of the mod suit while having a minigun as well

 I think this would be a nice TC trade for NukeOps or admin spawn for Death Squad Perhaps, maybe in the future it could be a Nuke Ops item
## Changelog
:cl:
add: add minigun modsuit module
balance: Add sounds to bringing out and putting away miniguns

/:cl:

Squishypone committed:
add minigun module

Squishypone committed:
Merge branch 'tgstation:master' into master

Squishypone committed:
Merge branch 'tgstation:master' into master

Squishypone committed:
Change the wording a little bit

Zonespace27 commented on code/modules/mod/modules/modules_antag.dm:
```diff
@@ -494,3 +494,39 @@
+	w_class = WEIGHT_CLASS_HUGE
+	idle_power_cost = DEFAULT_CHARGE_DRAIN * 0.5
+	use_power_cost = DEFAULT_CHARGE_DRAIN * 3
+	var/overheat = 0
+	var/overheat_max = 30
+	var/heat_diffusion = 0.5
```
move these below the vars that are inherited (not having a `var/` in front of them)

Zonespace27 commented on code/modules/mod/modules/modules_antag.dm:
```diff
@@ -494,3 +494,39 @@
 	if(deleting)
 		return
 	mod.helmet.flash_protect = initial(mod.helmet.flash_protect)
+/obj/item/mod/module/minigun
```
```suggestion

/obj/item/mod/module/minigun
```

Zonespace27 commented on code/modules/mod/modules/modules_antag.dm:
```diff
@@ -494,3 +494,39 @@
+	name = "MOD laser gatling gun module"
+	icon_state = "minigun_module"
+	inhand_icon_state = "minigun"
+	desc = "A laser gatling gun with a that connects to a MOD suit"
```
grammar is wrong here; put this above icon state vars but below name

Zonespace27 commented on code/modules/mod/modules/modules_antag.dm:
```diff
@@ -494,3 +494,39 @@
+	var/heat_diffusion = 0.5
+	module_type = MODULE_ACTIVE
+	complexity = 5
+	var/obj/item/stock_parts/cell/minigun/battery
```
same deal var-wise here with the others; also get rid of this on Destroy()

Zonespace27 commented on code/modules/mod/modules/modules_antag.dm:
```diff
@@ -494,3 +494,39 @@
+	. = ..()
+	battery = new(src)
+	
+	START_PROCESSING(SSobj, src)
```
```suggestion
	START_PROCESSING(SSobj, src)
	
```

Zonespace27 commented on code/modules/projectiles/guns/energy/laser_gatling.dm:
```diff
@@ -83,12 +84,15 @@
 /obj/item/minigunpack/update_icon_state()
 	icon_state = armed ? "notholstered" : "holstered"
 	return ..()
+/obj/item/minigunpack/proc/drain_power()
```
```suggestion

/obj/item/minigunpack/proc/drain_power()
```

Zonespace27 commented on code/modules/projectiles/guns/energy/laser_gatling.dm:
```diff
@@ -151,7 +167,7 @@
 
 /obj/item/gun/energy/minigun/afterattack(atom/target, mob/living/user, flag, params)
-	if(!ammo_pack || ammo_pack.loc != user)
+	if(!istype(ammo_pack, /obj/item/minigunpack)&&!istype(ammo_pack, /obj/item/mod/module/minigun/))
```
```suggestion
	if(!istype(ammo_pack, /obj/item/minigunpack) && !istype(ammo_pack, /obj/item/mod/module/minigun))
```

Zonespace27 commented on code/modules/projectiles/guns/energy/laser_gatling.dm:
```diff
@@ -115,12 +119,23 @@
-	AddElement(/datum/element/update_icon_blocker)
-	AddComponent(/datum/component/automatic_fire, 0.2 SECONDS)
-	return ..()
+	if(istype(loc, /obj/item/minigunpack)) 
+		ammo_pack = loc
+		AddElement(/datum/element/update_icon_blocker)
+		AddComponent(/datum/component/automatic_fire, 0.2 SECONDS)
+		return ..()
+	else if(istype(loc, /obj/item/mod/module/minigun))
+		ammo_pack = loc
+		AddElement(/datum/element/update_icon_blocker)
+		AddComponent(/datum/component/automatic_fire, 0.2 SECONDS)
+		return ..()
```
these if statements seem to do the same thing

Zonespace27 commented on code/modules/mod/modules/modules_antag.dm:
```diff
@@ -494,3 +494,39 @@
+
+/obj/item/mod/module/minigun/process(delta_time)
+	overheat = max(0, overheat - heat_diffusion * delta_time)
+	if(!overheat == 0)
```
```suggestion
	if(overheat)
```

Zonespace27 commented on code/modules/mod/modules/modules_antag.dm:
```diff
@@ -494,3 +494,39 @@
+/obj/item/mod/module/minigun/process(delta_time)
+	overheat = max(0, overheat - heat_diffusion * delta_time)
+	if(!overheat == 0)
+		drain_power(overheat*0.5)//It drains power if it overheats
```
```suggestion
		drain_power(overheat * 0.5)//It drains power if it overheats
```

Zonespace27 commented on code/modules/mod/modules/modules_antag.dm:
```diff
@@ -494,3 +494,39 @@
+	battery = new(src)
+	
+	START_PROCESSING(SSobj, src)
+/obj/item/mod/module/minigun/proc/attach_gun(mob/user)
```
code doc this
additionally, is this ever called? Not seeing it anywhere

Zonespace27 commented on code/modules/projectiles/guns/energy/laser_gatling.dm:
```diff
@@ -83,12 +84,15 @@
 /obj/item/minigunpack/update_icon_state()
 	icon_state = armed ? "notholstered" : "holstered"
 	return ..()
+/obj/item/minigunpack/proc/drain_power()
```
i'm like. 99% sure this isn't needed

Squishypone replied:
It will refuse to compile without it but if it's really that bad I could find a way around doing it this way

Zonespace27 reviewed (changes requested):
You'll want to fix the iconstate CI failures too

Zonespace27 commented on code/modules/projectiles/guns/energy/laser_gatling.dm:
```diff
@@ -115,12 +119,23 @@
 	var/obj/item/minigunpack/ammo_pack
```
Ahh, i see what you're doing here

This is bad practice, consider componentizing this instead

Squishypone replied:
I guess I could just make it so it detects whether or not it's in a mod suit or not I'll try that

Squishypone replied:
Not sure I'm going to start with this I'm really new to this stuff, so I don't really know how to do that because that was just the code there before I started
Is there any examples of something similar like this done correctly?

Zonespace27 replied:
https://hackmd.io/@tgstation/SignalsComponentsElements

Ping me in the code channel if you're not sure after reading this

Squishypone committed:
a bunch of suggestions fix

Squishypone committed:
Merge branch 'master' into master

Squishypone committed:
Fix the in hand thing, I think

Squishypone committed:
Merge branch 'master' of https://github.com/Squishypone/tgstation

Rex9001 commented:
Doesnt this go directly against the modsuit design doc?

CRITAWAKETS commented:
i'd recommend adding a sprite for the minigun module using the already existing module overlay sprites, that'd look _really_ cool

Squishypone commented:
> i'd recommend adding a sprite for the minigun module using the already existing module overlay sprites, that'd look _really_ cool
Which sprite are you talking about?

Squishypone committed:
Change stuff around so we don't need useless procs

I removed the useless procs and made a way for it to work without it and some other minor changes, that should be fine

Squishypone committed:
Fixed sound error

Squishypone committed:
one space

Squishypone committed:
A tiny bit of formatting

JohnFulpWillard reviewed (commented):
Marked as draft to deal with reviews

I'm not a fan of ``attach_gun`` at all and don't like how you're basically making a reference to 2 different things that they can only be connected to 1 of at any time. Please consider making a component for it.

Squishypone committed:
implement components and change the shot cost system

Squishypone committed:
Merge branch 'master' into master

Squishypone commented:
I think this is what you wanted with the components. Can we take a look and tell me what you think?

Squishypone commented:
oh thanks I'm new the GitHub I'll do that

JohnFulpWillard reviewed (changes requested):
The minigun pack and minigun module should have a component that allows it to extend out the gun on an attack hand signal.

The component should be listening to the minigun's signals such as ``COMSIG_GUN_FIRED`` for shooting, ``COMSIG_ITEM_AFTERATTACK`` for afterattack, etc.
The minigun itself should preferably be doing nothing but adding itself their update_icon_blocker element and automatic_fire component.

Squishypone commented:
> The minigun pack and minigun module should have a component that allows it to extend out the gun on an attack hand signal.
> 
> The component should be listening to the minigun's signals such as `COMSIG_GUN_FIRED` for shooting, `COMSIG_ITEM_AFTERATTACK` for afterattack, etc. The minigun itself should preferably be doing nothing but adding itself their update_icon_blocker element and automatic_fire component.

 So you're saying I should move all the processing from the gun to the pack and module as a component?  Also, I don't know how that would affect the mini gun as the mini gun takes power from the cell that is inside the pack.  Thank you for the feedback. I'll try to figure out this.

Squishypone committed:
Merge branch 'master' into master

JohnFulpWillard reviewed (commented):
Posting here since you asked on the Discord;

The idea is you'd have a ``/datum/component`` that registers to being attack hand, that's what you'd use to take the gun in/out. ``overheat`` can be moved to the component entirely, while the power problem can be fixed with a callback the component will invoke when shot.

sorta like this
```
/datum/component/minigun
    var/obj/item/gun/gun //this is moved from the minigun to being here.
    var/datum/callback/on_firing_callback //you will call this when its shot

/datum/component/minigun/Initialize(on_firing_callback)
   if(!istype(parent, /obj/item/mod/module/minigun) && !istype(parent, /obj/item/minigunpack))
        return COMPONENT_INCOMPATIBLE
    src.on_firing_callback = on_firing_callback //this will invoke when you call Invoke() on it, set it to call when you need to. You should also check that it exists when you do, in case someone puts this component on something they don't want to have a callback.
    gun = new gun(src)
    RegisterSignal(gun, SIGNAL_SENT_BY_GUNS_WHEN_FIRING, PROC_REF(on_gun_fire))
    RegisterSignal(parent, SIGNAL_SENT_WHEN_CLICKING_WITH_HAND, PROC_REF(on_attack_hand))
```

from there you can make the ``on_gun_fire`` proc do the effects you want when the gun is fired, and ``on_attack_hand`` be what lets you take the gun out. You can register more signals to the other stuff you wanna do, like putting it back in via clicking on it with the gun or dropping it, refilling, overheating, etc.
